Fix incorrect custom pseudo-scrollbar sizes caused by scroller style changes on Mac.

GetToSet opened the pull request:
#### 560abdc4b29416fc6052fa10fafd84f00c7a4244
<pre>
Fix incorrect custom pseudo-scrollbar sizes caused by scroller style changes on Mac.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249892">https://bugs.webkit.org/show_bug.cgi?id=249892</a>

Reviewed by Simon Fraser.

Scroller style changes can cause custom pseudo-scrollbar sizes (set by
`width` or `height` property using `::-webkit-scrollbar`) to be
overridden by AppKit defaults. This could cause unexpected visual
appearances, especially when the scroller is thinner than the defaults.

This usually happens when the user enters or leaves the trackpad-only
setup (connecting or disconnecting the mouse for laptops would be a
typical case). Manually toggling the scroller appearance in the system
settings would be another way to reproduce this issue.

This patch resolves the above issue by explicitly checking for custom
scrollbars before replacing them with standard implementations when
processing scroller style updates.

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/scrollbars/custom-scrollbar-scroller-style-change-expected.txt: Added.
* LayoutTests/scrollbars/custom-scrollbar-scroller-style-change.html: Added.
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setUsesOverlayScrollbars):
* Source/WebCore/testing/Internals.mm:
(WebCore::Internals::setUsesOverlayScrollbars):

Canonical link: <a href="https://commits.webkit.org/259389@main">https://commits.webkit.org/259389@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c1046546c07b1be873582aa23076a75b9edac0f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11272 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111454 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171629 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106110 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12251 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2186 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94509 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109191 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107907 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9381 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92658 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37188 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91250 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24138 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78914 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4827 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25563 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4941 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2000 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10996 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45059 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6476 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6686 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/scrollbars/custom-scrollbar-scroller-style-change-expected.txt:
```diff
@@ -0,0 +1,9 @@
+layer at (0,0) size 800x600
```
Does this render tree dump sufficiently test the bug? If so, there's no need to add the PNG file. If not, it should be converted to a reference test.

GetToSet replied:
The render tree dump would be enough for this issue (both wk1 & 2 verified). Could you please further explain the reference test since I couldn't find any materials about this.

smfr replied:
A reference test uses two html files: `foo.html`, and `foo-expected.html` which are expected to render identically. You construct `foo-expected.html` to not exercise the buggy code, but give the same pixel results.

GetToSet committed:
Fix incorrect custom pseudo-scrollbar sizes caused by scroller style changes on Mac.
https://bugs.webkit.org/show_bug.cgi?id=249892

Reviewed by Simon Fraser.

Scroller style changes can cause custom pseudo-scrollbar sizes (set by
`width` or `height` property using `::-webkit-scrollbar`) to be
overridden by AppKit defaults. This could cause unexpected visual
appearances, especially when the scroller is thinner than the defaults.

This usually happens when the user enters or leaves the trackpad-only
setup (connecting or disconnecting the mouse for laptops would be a
typical case). Manually toggling the scroller appearance in the system
settings would be another way to reproduce this issue.

This patch resolves the above issue by explicitly checking for custom
scrollbars before replacing them with standard implementations when
processing scroller style updates.

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/scrollbars/custom-scrollbar-scroller-style-change-expected.txt: Added.
* LayoutTests/scrollbars/custom-scrollbar-scroller-style-change.html: Added.
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setUsesOverlayScrollbars):
* Source/WebCore/testing/Internals.mm:
(WebCore::Internals::setUsesOverlayScrollbars):

Canonical link: https://commits.webkit.org/259389@main
